Ashrita - Job Posting Page Analytics: Create a Country of Application Map Chart

AshritaCherlapally opened the pull request:
# Description
<img width="697" height="513" alt="image" src="https://github.com/user-attachments/assets/8f42c9ae-45f9-4091-80e9-b9bca7600ad3" />

## Related PRS (if any):
To test this backend PR you need to checkout the #XXX frontend PR.
…

## Main changes explained:
- New endpoints (mounted under /applications)
- GET /applications/roles — returns distinct roles (sorted).
- GET /applications/applications — aggregates applicants by country for the selected timeframe and optional roles.
- GET /applications/comparison — returns per-country % change and overall % vs previous period.

## How to test:
1. check into current branch
2. do `npm install` and `npm start` to run this PR locally
3. Send a POST request to  http://localhost:4500/api/login to obtain a token
4. GET http://localhost:4500/applications/roles to get all the roles listed.
5. GET http://localhost:4500/applications/applications?filter=monthly to filter out the applications based on monthly filter.
6. GET http://localhost:4500/applications/applications?filter=monthly&roles=Engineer,Designer to filter out the roles with Monthly filter.
7. GET http://localhost:4500/applications/comparison?filter=monthly to compare the monthly applications.


## Screenshots or videos of changes:
<img width="1050" height="809" alt="image" src="https://github.com/user-attachments/assets/347ac0ba-d9fa-49e2-8cb1-00411de34ae6" />
<img width="1053" height="809" alt="image" src="https://github.com/user-attachments/assets/cb6d1f26-db71-45bf-ad7e-a9ba034a02cc" />
<img width="1053" height="809" alt="image" src="https://github.com/user-attachments/assets/3feeb4ba-a36e-442b-987e-71fa2e250e13" />
<img width="1057" height="815" alt="image" src="https://github.com/user-attachments/assets/8611558f-d853-459c-af6b-0914fd3cf8dc" />

Aditya-gam reviewed (changes requested):
- Unsure about the correct endpoints. Should they be `localhost:4500/api/application` or `localhost:4500/applications`?
- GET [http://localhost:4500/applications/roles](http://localhost:4500/applications/roles) successfully returns an array of all available roles.
    - <img width="1127" height="893" alt="GetAllRolesSuccess" src="https://github.com/user-attachments/assets/1642cc9d-5a59-4ff4-88b6-9cd069ae872a" />

- Tested aggregation by country with weekly, monthly, and yearly filters:
    - Weekly and monthly filters return empty arrays.
    - Yearly filter returns valid data in the response.
    - <img width="1127" height="893" alt="AggByCountryFilterYearly" src="https://github.com/user-attachments/assets/9a3fe637-c104-4c17-bcaa-3e91ffa4e297" />

- Compared applications across monthly and yearly filters:
    - Both return valid responses with data.
    - Weekly filter returns valid response but with empty data.
    - <img width="1127" height="893" alt="CompareMonthlyApplications" src="https://github.com/user-attachments/assets/7786111c-f44a-471c-89ef-27b8b1d838ae" />

- Request without a filter on `/applications` (`http://localhost:4500/applications/applications`):
    - Returns a valid response with an empty array, defaulting to the monthly filter.
    - Expected behavior might instead be an error (400 or 422).
    - <img width="1127" height="893" alt="MissingFilterOnApplications" src="https://github.com/user-attachments/assets/eb061390-3c79-411e-acc2-f3f48ffa7177" />

- Invalid filters on `/applications` (e.g., `daily`) return a valid response instead of an error.
    - Expected behavior might be to return 400 or 422, since only weekly, monthly, and yearly are supported.
    - <img width="1127" height="893" alt="InvalidFilterOnApplication" src="https://github.com/user-attachments/assets/1dd86672-9eab-480f-bf74-0e7147f98240" />

- Filtering by unknown roles returns a valid response with an empty array, which seems correct.
    - <img width="1127" height="893" alt="UnknownRolesFilter" src="https://github.com/user-attachments/assets/9fb8db03-2056-4ccc-8ca9-c3037715e8ee" />

- Request without a filter on `/comparison` (`http://localhost:4500/applications/comparison`):
    - Returns a valid response with an empty array, defaulting to the monthly filter.
    - Expected behavior might instead be a 400 or 422 error.
    - <img width="1127" height="893" alt="MissingFilterComparisons" src="https://github.com/user-attachments/assets/a271d11b-343f-4796-ac57-45c0af7ba2b7" />

- Invalid filters on `/comparison` (e.g., `quarterly`) return a valid response instead of an error.
	- Expected behavior might be to return 400 or 422.
	- <img width="1127" height="893" alt="InvalidFilterOnComparisons" src="https://github.com/user-attachments/assets/acdb310e-133e-4dc5-8412-bf47aad95b46" />

- Ran concurrent functional and performance tests.
	- https://github.com/user-attachments/assets/542284c2-6398-487c-aa8e-42f1c7351b61

AshritaCherlapally commented:
Addressed the feedback and made code changes accordingly:
Summary
* Implemented Applications API (roles, aggregation, comparison) with validation & caching.
* Removed duplicate files and unified model/route names.
Endpoints
* GET /applications/roles → sorted list of roles.
<img width="1056" height="812" alt="Screenshot 2025-09-19 at 6 49 41 PM" src="https://github.com/user-attachments/assets/c062c4e8-b19b-4682-9b8d-70ecb0140d89" />
* GET /applications?filter=weekly|monthly|yearly[&roles=a,b]→ aggregate by country for preset range.
<img width="1057" height="808" alt="Screenshot 2025-09-19 at 6 51 24 PM" src="https://github.com/user-attachments/assets/30278f9f-5718-43d6-9a39-2c5fff2977c9" />
<img width="1054" height="812" alt="Screenshot 2025-09-19 at 6 52 18 PM" src="https://github.com/user-attachments/assets/490b6f1b-0e5e-4c8d-9169-ef2b98f20b11" />
<img width="1059" height="811" alt="Screenshot 2025-09-19 at 6 52 38 PM" src="https://github.com/user-attachments/assets/be650f1b-9acb-40a4-9d6a-b5e3bc92d8eb" />
<img width="1059" height="811" alt="Screenshot 2025-09-19 at 6 53 15 PM" src="https://github.com/user-attachments/assets/229bd9c9-cbbd-4c6e-abbf-22766c44d473" />
* GET /applications?startDate=ISO&endDate=ISO[&roles=a,b] → custom range (aggregation only).
* GET /applications/comparison?filter=weekly|monthly|yearly[&roles=a,b] → % change vs previous period.
<img width="884" height="714" alt="Screenshot 2025-09-19 at 8 13 59 PM" src="https://github.com/user-attachments/assets/df12b236-8ddb-4b9f-a38b-d9315a21550f" />
<img width="893" height="709" alt="Screenshot 2025-09-19 at 8 14 14 PM" src="https://github.com/user-attachments/assets/72952e30-db1a-4dc4-8d34-ca2b79a77bd4" />
<img width="898" height="716" alt="Screenshot 2025-09-19 at 8 14 26 PM" src="https://github.com/user-attachments/assets/919405f0-69a9-41e3-bbf0-759d7b090696" />
Validation (per feedback)
* Missing filter on /applications and /applications/comparison (and no custom dates) → 422.
<img width="890" height="711" alt="Screenshot 2025-09-19 at 8 15 18 PM" src="https://github.com/user-attachments/assets/64e0a781-63ed-45ef-b6b8-0caaa4185f63" />
<img width="892" height="706" alt="Screenshot 2025-09-19 at 8 15 36 PM" src="https://github.com/user-attachments/assets/b3a384be-3721-431f-b269-fa03176c977f" />
<img width="904" height="722" alt="Screenshot 2025-09-19 at 8 15 54 PM" src="https://github.com/user-attachments/assets/6dfdfbf3-96b0-4378-9f20-f74dbb59c54d" />
* Invalid filter (e.g., daily, quarterly) → 422.
<img width="889" height="703" alt="Screenshot 2025-09-19 at 8 16 13 PM" src="https://github.com/user-attachments/assets/0a9e41f0-c1f3-4cbd-a61f-bb53b0eac25f" />

Prem203 commented:
I reviewed the PR and I am getting 404 error for all the endpoints. Let me know if there is any specific problem or if I am missing something. Below are the screenshots for the same:

<img width="1387" height="871" alt="image" src="https://github.com/user-attachments/assets/aa115d9b-ca7e-4082-968b-3d055ae798d5" />

<img width="1385" height="859" alt="image" src="https://github.com/user-attachments/assets/a8aa6623-2c2f-4be5-a419-d1c35ce3c8f1" />

<img width="1386" height="854" alt="image" src="https://github.com/user-attachments/assets/d54fe5a7-6876-498e-a143-90f4870f5ea1" />

<img width="1390" height="862" alt="image" src="https://github.com/user-attachments/assets/cfb41e43-2905-490a-b6eb-87dddec00d26" />

one-community commented:
Thank you all, merging!
